File: examples/python/qft.py
import qristal.core
import numpy as np
import multiprocessing

nb_qubits = 5
nb_shots = 1000
circ = qristal.core.Circuit()

# Apply the QFT circuit
circ.qft(range(nb_qubits))

# Apply measurement
circ.measure_all()

# Print out circuit
circ.print()

# List of backends
backends = ["aer_mps", "aer_dm", "tnqvm", "cudaq:dm"]

# Create list of sessions to be executed in parallel
sessions = []
for backend in backends:
    s = qristal.core.session()
    s.init()
    s.qn = nb_qubits
    s.sn = nb_shots
    s.ir_target = circ

    if backend == "aer_mps":
        s.acc = "aer"
        s.aer_sim_type = "matrix_product_state"
        s.aer_omp_threads = 4
    elif backend == "aer_dm":
        s.acc = "aer"
        s.aer_sim_type = "density_matrix"
        s.aer_omp_threads = 4
    else:
        s.acc = backend
    sessions.append(s)

# Execute circuit and retrieve distribution
def get_prob_dist(sim, results):
    s.run()
    measured_probs_dict = {}
    for i in s.results[0][0]:
        measured_probs_dict[str(i)[::-1]] = s.results[0][0][i] / s.results[0][0].total_counts()
    results.append(measured_probs_dict)

# Execute backends in parallel
processes = []
manager = multiprocessing.Manager()
all_results = manager.list()  # This will be shared across all processes
for sim in sessions:
    p = multiprocessing.Process(
        target = get_prob_dist,
        args = (sim, all_results)
    )
    processes.append(p)
    p.start()

# Wait for all workers to complete task
for p in processes:
    p.join()

# Print all results
for i in np.arange(len(backends)):
    print("Results", backends[i], ":", all_results[i], "\n")

File: examples/cpp/qft/qft.cpp
// Copyright (c) Quantum Brilliance Pty Ltd

#include <qristal/core/session.hpp>
#include <qristal/core/circuit_builder.hpp>
#include <qristal/core/thread_pool.hpp>

#include <string>
#include <iostream>
#include <future>
#include <chrono>
#include <thread>

// Add job handler
std::string run_async(qristal::session& s) {
  s.run();
  std::ostringstream out;
  for (const auto& [bits, count] : s.results()) {
    for (size_t i = bits.size(); i-- > 0; ) out << bits[i];
    out << ": " << count << std::endl;
  }
  return out.str();
};

int main()
{
  constexpr int num_qubits = 5;
  constexpr int num_shots = 1024;
  constexpr int aer_instance_thread_limit = 4;

  // Create QFT circuit
  qristal::CircuitBuilder circ;
  std::vector<int> qft_qubits(num_qubits);
  // Fill the qubit list with 0, 1, ..., n-1
  // i.e., the qubits that we want to apply the QFT circuit to.
  std::iota(std::begin(qft_qubits), std::end(qft_qubits), 0);
  // Apply QFT
  circ.QFT(qft_qubits);
  circ.MeasureAll(num_qubits);

  // Make four Qristal sessions corresponding to the 4 backends we want to run: aer:matrix_product state, aer:density_matrix, cudaq:dm & tnqvm
  constexpr size_t num_jobs = 4;
  std::vector<std::future<std::string>> futures{};
  std::vector<qristal::session> my_sims(num_jobs);
  my_sims[0].acc = "aer";
  my_sims[0].aer_sim_type = "matrix_product_state";
  my_sims[1].acc = "aer";
  my_sims[1].aer_sim_type = "density_matrix";
  my_sims[2].acc = "cudaq:dm";
  my_sims[3].acc = "tnqvm";
  for (auto& sim : my_sims) {
    sim.gpu_device_ids = {0};
    sim.qn = num_qubits;
    sim.sn = num_shots;
    sim.aer_omp_threads = aer_instance_thread_limit;
    sim.irtarget = circ.get();
     // Run async jobs
    futures.push_back(qristal::thread_pool::submit(run_async, std::ref(sim)));
    std::this_thread::sleep_for(std::chrono::milliseconds(1));
  }

  std::future_status status;
  std::vector<bool> finished(num_jobs, false);
  std::size_t counter{};
  std::size_t loopCounter{0};
  do {
    std::this_thread::sleep_for(std::chrono::milliseconds(1000));
    counter = 0;
    std::stringstream msg;

    for (size_t i = 0; i < num_jobs; i++) {
      if (finished[i] == false) {
        status = futures[i].wait_for(std::chrono::milliseconds(0));
        if (status == std::future_status::ready){
          finished[i] = true;
        }
      }

      if (finished[i] == true) {
        counter++;
      }
    }
    loopCounter++;
  } while (counter < num_jobs);

  // Retrieve results from futures
  std::vector<std::string> results{};
  for (size_t i = 0; i < num_jobs; i++) {
    results.push_back(futures[i].get());
    std::cout << "Results[" << i << "]:\n" << results[i] << std::endl;
    if (results[i].empty()) {
      std::cout << "Results[" << i << "] is empty!" << std::endl;
    }
  }
}
